Meta/Lagom/Tools/CodeGenerators/LibUnicode/GenerateIDNAData.cpp

```diff
@@ -0,0 +1,240 @@
+    Ignored,
+    Mapped,
+    Deviation,
+    Disallowed,
+    DisallowedStd3Valid,
+    DisallowedStd3Mapped,
+};
+
+static constexpr Array<StringView, 7> mapping_status_names { "Valid"sv, "Ignored"sv, "Mapped"sv, "Deviation"sv, "Disallowed"sv, "DisallowedStd3Valid"sv, "DisallowedStd3Mapped"sv };
+
+enum class IDNA2008Status : u8 {
+    NV8,
+    XV8,
+};
+
+static constexpr Array<StringView, 2> idna_2008_status_names { "NV8"sv, "XV8"sv };
+
+struct IDNAMapping {
+    Unicode::CodePointRange code_points;
+    MappingStatus status;
+    IDNA2008Status idna_2008_status;
+    Vector<u32> mapped_to {};
+};
+
+struct IDNAData {
+    Vector<IDNAMapping> mapping_table;
+};
+
+static MappingStatus parse_mapping_status(StringView status)
+{
+    if (status == "valid"sv)
+        return MappingStatus::Valid;
+    if (status == "ignored"sv)
+        return MappingStatus::Ignored;
+    if (status == "mapped"sv)
+        return MappingStatus::Mapped;
+    if (status == "deviation"sv)
+        return MappingStatus::Deviation;
+    if (status == "disallowed"sv)
+        return MappingStatus::Disallowed;
+    if (status == "disallowed_STD3_valid"sv)
+        return MappingStatus::DisallowedStd3Valid;
+    if (status == "disallowed_STD3_mapped"sv)
+        return MappingStatus::DisallowedStd3Mapped;
+    VERIFY_NOT_REACHED();
+}
+
+static ErrorOr<void> parse_idna_mapping_table(Core::InputBufferedFile& file, Vector<IDNAMapping>& mapping_table)
+{
+    Array<u8, 1024> buffer;
+
+    while (TRY(file.can_read_line())) {
+        auto line = TRY(file.read_line(buffer));
+
+        if (line.is_empty() || line.starts_with('#'))
+            continue;
+
+        if (auto index = line.find('#'); index.has_value())
+            line = line.substring_view(0, *index);
+
+        auto segments = line.split_view(';', SplitBehavior::KeepEmpty);
+        VERIFY(segments.size() >= 2);
+
+        IDNAMapping idna_mapping {};
+        idna_mapping.code_points = parse_code_point_range(segments[0].trim_whitespace());
+        idna_mapping.status = parse_mapping_status(segments[1].trim_whitespace());
+
+        if (segments.size() >= 3)
+            idna_mapping.mapped_to = parse_code_point_list(segments[2].trim_whitespace());
+
+        if (segments.size() >= 4) {
+            auto trimmed = segments[3].trim_whitespace();
+            if (trimmed == "NV8"sv) {
+                idna_mapping.idna_2008_status = IDNA2008Status::NV8;
+            } else {
+                VERIFY(trimmed == "XV8"sv);
+                idna_mapping.idna_2008_status = IDNA2008Status::XV8;
+            }
+        }
+
+        TRY(mapping_table.try_append(move(idna_mapping)));
+    }
+
+    return {};
+}
+
+static ErrorOr<void> generate_idna_data_header(Core::InputBufferedFile& file, IDNAData&)
+{
+    StringBuilder builder;
+    SourceGenerator generator { builder };
+
+    generator.append(R"~~~(
+#pragma once
+
+#include <AK/Optional.h>
+#include <AK/Utf32View.h>
+#include <LibUnicode/CharacterTypes.h>
```

Comment:
These are unused in this header (if they're needed in the .cpp impl, let's move them there).